### DISCUSSION TITLE
My Fastapi application is not able to scale above 1000m CPU in kubernetes pod

### DISCUSSION THREAD

**Discussion by Ayusharma0698** [fastapi/fastapi#9585] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app = FastAPI()
SERVICE_NAME = "My app"


def init_service(host: str, port: int, log_level, tier: str):
    app.include_router(health_check_controller.service_router)
    app.include_router(parse_file_controller.service_router)
    app.title = SERVICE_NAME
    uvicorn_server = setup_logging(app, host, port, tier, log_level)
    return uvicorn_server


def setup_logging(application: SERVICE_NAME, host: str, port: int, tier: str, level):
    uvicorn_server = Server(
        Config(
            app=application,
            host=host,
            port=port,
            log_level=level,
            workers=4
        )
    )
    # some logging setup
    return uvicorn_server


if __name__ == "__main__":
    init_service('0.0.0.0', 8080, logging.INFO, 'dev').run()
```

### Description

I have a python application using fastapi and uvicorn that has some heavy cpu usage tasks. Now while running this on a kubernetes pod the application is not able to go above 1000m CPU usage. The following configuration is given for the pod.

`resources:
  limits:
    cpu: "4"
    memory: 5000Mi
  requests:
    cpu: "2"
    memory: 2000Mi
`

I am doing load testing using k6. The application is also giving a 504 response Gateway timed out.

The following is the k6 script.js used for load testing 

`
import http from 'k6/http';

import { check, sleep } from 'k6';

export const options = {

  stages: [

    { duration: '300s', target: 1000 },

    { duration: '150s', target: 500 },

    { duration: '10s', target: 100 },

  ],

};

export default function () {

  const res = http.get('http://0.0.0.0:8080');

  check(res, { 'status was 200': (r) => r.status == 200 });

  sleep(1);

}
`
The application is build to a docker image and deployed in the kubernetes pod. 

Please help me in finding out why the application is not able to utilise atleast 2 core of CPU.

### Operating System

Linux, macOS

### FastAPI Version

0.89.1

### Python Version

3.11

**Comment by jgould22** [comment]
Uvicorn defaults to one worker it will not use multiple cores just because it is there.

You can see the settings here https://www.uvicorn.org/deployment/

Try setting WEB_CONCURRENCY environment var and see if that fixes it. You can also pass a param to your run function.

Additionally you need to be aware that you should not be doing CPU bound tasks in an asynchronous this will degrade application performance.

**Reply by Ayusharma0698** [reply]
Hi @jgould22 ,

I have given my unicorn workers as 4 already. 
The following is what my code looks like from main.py

`app = FastAPI()
SERVICE_NAME = "My app"


def init_service(host: str, port: int, log_level, tier: str):
    app.include_router(health_check_controller.service_router)
    app.include_router(parse_file_controller.service_router)
    app.title = SERVICE_NAME
    uvicorn_server = setup_logging(app, host, port, tier, log_level)
    return uvicorn_server


def setup_logging(application: SERVICE_NAME, host: str, port: int, tier: str, level):
    uvicorn_server = Server(
        Config(
            app=application,
            host=host,
            port=port,
            log_level=level,
            workers=4
        )
    )
    # some logging setup
    return uvicorn_server


if __name__ == "__main__":
    init_service('0.0.0.0', 8080, logging.INFO, 'dev').run()`

**Reply by jgould22** [reply]
My apologies I missed the workers definition in the config.

What are the endpoints in your application doing?, are you running blocking code in an async  endpoint?

**Reply by Ayusharma0698** [reply]
My endpoints are also defined as def. I checked my code that if its blocking a thread or not. I was able to see it occupying different threads.
There are different endpoints in my app. The major one where scaling issue is there is the excel to csv end point. I am using pandas for excel to csv conversion.

But if I leave the endpoints aside even pinging the application through a load tester is not taking it above 1000m CPUs. Is unicorn so restrictive?

**Comment by DanielElisenberg** [comment]
I would recommend this write-up by @tiangolo
https://fastapi.tiangolo.com/deployment/server-workers/
as it might shed some light on your issue. Uvicorn is not a great process manager, so when we want to utilize more than one cpu, gunicorn with uvicorn workers might be the way to go. (I have not tried this myself)

As a side note; Have you tried letting k8s handle this by replicating the pod?

**Reply by Ayusharma0698** [reply]
I am not able to find a great working sample of gunicorn using docker image and kubernetecs. Using gunicorn I am able to start my application in a local environment but was not able to scale up to a docker deployment.

For the pod replication, I have not set the autoscaling to be true as I was not able to utlise my config of even one pod.

My k8 config looks something like this
`autoscaling:
  enabled: false
  minReplicas: 1
  maxReplicas: 3
  targetCPUUtilizationPercentage: 80
  targetMemoryUtilizationPercentage: 80`

**Comment by Kludex** [comment]
The 4 workers are not being used in your code. I'll provide details later.

**Reply by Ayusharma0698** [reply]
Please provide the details.

**Reply by Kludex** [reply, marked as answer]
The spawn of processes doesn't happen on the `Server` object. It's here: https://github.com/encode/uvicorn/blob/f764e503971cbf5b037dd324f2add652e602adc4/uvicorn/main.py#L574-L576

**Reply by Ayusharma0698** [reply]
Can you suggest an alternative approach?

**Reply by Kludex** [reply]
I used this in the past: https://gist.github.com/Kludex/c98ed6b06f5c0f89fd78dd75ef58b424

**Reply by Ayusharma0698** [reply]
Thank you for the help this works.
